outline/concurrency.md

```diff
@@ -0,0 +1,84 @@
+Running tasks concurrently
+
+# Pages:
+
+* [Overview](#overview)
+* [Channels](#channels)
+* [Backpressure](#backpressure)
+
+<a name="overview"></a>
+## Overview
+
+**Status**: Unassigned.
+
+Overview of concurrency and parallism with TOkio.
+
+#### Contents
+
+* Concurrency in an asynchronous system.
+  * Asynchronous gets you the ability to run concurrent tasks.
+  * Concurrency is not parallism.
+    * Parallism requires running on multiple cores.
+  * Parallism is required to scale on modern servers
+    * Modern servers have many cores.
+* Tokio offers two runtimes, one parallel, one not.
+  * `tokio::run`
+    * Concurrent and parrallel (if the system supports it).
+    * Runs multiple OS threads.
+    * Tokio tasks are multiplexed across multiple threads.
+      * M:N scheduling.
+  * `tokio::runtime::current_thread`
+    * Concurrent, but **not** parrallel.
+    * Single OS thread.
+  * Pros / cons to both
+  * Uses Rust to enforce thread-safety.
+    * Tasks spawned on `tokio::run` must be `Send`.
+    * Tasks spawned on `runtime::current_thread` must not be `Send`.
+
+<a name="channels"></a>
+## Channels
+
+**Status**: Unassigned.
+
+Message passing for managing concurrency. This covers both oneshot and mpsc.
+
+#### Contents
+
+* Concurrency requires coordination
+  * Thread-safe as well (when parrallel).
+  * Many coordination/synchronization primitives that work in "synchronous"
+    modes of programming do not carry over to asynchronous.
+    * Mutexes don't work -- not possible to implement a future aware Mutex.
```

Comment:
What do you mean by that? Sure it is possible to implement a future aware Mutex, no?

Reply:
@stjepang It's the generalized multi consumer problem. One cannot implement a future aware mutex for the same reason that one cannot implement a multi consumer channel.

There are ways of working around... but they runtime opt-in. This could be rephrased to something more like "future aware mutexes do not exist today. They are not the correct way to manage synchronization with Tokio" instead...